src/file_pipe_log/pipe.rs

```diff
@@ -476,6 +477,46 @@ impl<F: FileSystem> SinglePipe<F> {
         self.flush_metrics(current.total_len);
         Ok((current.first_seq_in_use - prev.first_seq_in_use) as usize)
     }
+
+    fn snapshot(&self) -> (FileSeq, usize) {
+        let active_file = self.active_file.lock();
+        (active_file.seq, active_file.writer.offset())
+    }
+
+    fn restore(&self, file_seq: FileSeq, offset: usize) -> Result<()> {
+        let mut active_file = self.active_file.lock();
+        let mut files = self.files.write();
+        if file_seq < files.first_seq_in_use {
+            return Err(box_err!("Cannot truncate all files"));
+        }
+        while active_file.seq > file_seq {
+            assert!(files.fds.len() > 1);
+            let new_active_fd = &files.fds[files.fds.len() - 2];
+            let new_active_file = ActiveFile {
+                seq: active_file.seq - 1,
+                writer: build_file_writer(
+                    self.file_system.as_ref(),
+                    new_active_fd.handle.clone(),
+                    new_active_fd.format,
+                    false, /* force_reset */
+                )?,
+                format: new_active_fd.format,
+            };
+            let file_id = FileId {
+                queue: self.queue,
+                seq: active_file.seq,
+            };
+            self.file_system
+                .delete(file_id.build_file_path(&self.dir))?;
+            self.sync_dir()?;
```

Comment:
Why not sync after all are deleted?

Reply:
If there's a hole between files, recovery will discard older files. Here we must make sure each file is deleted before moving forward, to avoid leaving hole.

Reply:
If I delete a, and then delete b in the same directory, is it possible that b is actually deleted while a still exists? On the other hand, maybe recovery should keep deleting files even some are missing instead of breaking the looop.

Reply:
I believe it is possible.
> should keep deleting files even some are missing instead of breaking the looop.

It will cause data loss. Assuming b(seq=10) is deleted but a(seq=11) is not, on next reboot all files with seq<11 will be discarded.